File: frontend/api/corpus.ts
/**
 * 语料库 API 客户端
 */

const API_BASE_URL = 'http://localhost:8000/api';

// 类型定义
export interface ScenarioTag {
  label: string;
  type: string;
}

export interface CorpusItem {
  id: number;
  name: string;
  sentences: string;
  sTok: string;
  tTok: string;
  tags: ScenarioTag[];
  domain: string;
  source_type: string;
  is_public: boolean;
}

export interface CorpusListResponse {
  items: CorpusItem[];
  total: number;
  page: number;
  limit: number;
}

export interface BasicLayer {
  sentence_id: string;
  timestamp: string | null;
  platform: string | null;
}

export interface LanguageLayer {
  source_text_zh: string;
  raw_text_ms: string;
  normalized_text_ms: string;
  english_loanwords: string[];
}

export interface PragmaticLayer {
  intent: string[];
  sentiment: string;
  business_scenario: string | null;
}

export interface StyleLayer {
  style: string | null;
  contains_rojak: boolean;
  abbreviations_handled: Record<string, string>;
}

export interface CorpusSample {
  basic_layer: BasicLayer;
  language_layer: LanguageLayer;
  pragmatic_layer: PragmaticLayer;
  style_layer: StyleLayer;
}

export interface SampleListResponse {
  items: CorpusSample[];
  total: number;
  page: number;
  limit: number;
}

export interface CategoryStat {
  name: string;
  sentences: string;
  percent: string;
}

export interface DashboardOverviewResponse {
  corpus_count: string;
  total_pairs: string;
  language_count: string;
  categories: CategoryStat[];
}

export interface DashboardStatsResponse {
  total_sentences: string;
  sentences_growth: string;
  avg_quality_score: string;
  quality_target: string;
  localization_value: string;
  timeline_data: { month: string; count: number }[];
  intents: { label: string; val: number; w: string }[];
  sentiment_distribution: Record<string, number>;
  channel_sentiment: { channel: string; positive: number; negative: number; score: number }[];
  samples: CorpusSample[];
  loanwords: { word: string; count: number }[];
  normalization_map: { source: string; target: string; count: number }[];
  style_distribution: { scenario: string; colloquial: number; formal: number }[];
}

interface ApiResponse<T> {
  success: boolean;
  message: string;
  data: T;
  error?: string;
}

// 通用请求函数
async function apiRequest<T>(endpoint: string, options: RequestInit = {}): Promise<T> {
  const url = `${API_BASE_URL}${endpoint}`;

  // 获取 token
  const token = localStorage.getItem('auth_token');

  const headers: HeadersInit = {
    'Content-Type': 'application/json',
    ...options.headers,
  };

  if (token) {
    (headers as Record<string, string>)['Authorization'] = `Bearer ${token}`;
  }

  let response;
  try {
    response = await fetch(url, {
      ...options,
      headers,
    });
  } catch (err) {
    // 网络错误
    throw new Error('无权访问此语料库');
  }

  if (!response.ok) {
    // 尝试读取后端返回的错误信息
    try {
      const errorResult = await response.json();
      const errorMessage = errorResult.detail || errorResult.message || response.statusText;
      throw new Error(errorMessage);
    } catch (err) {
      // 如果已经是我们抛出的 Error，直接重新抛出
      if (err instanceof Error && (err.message === '无权访问此语料库' || err.message.includes('403'))) {
        throw err;
      }
      // 其他情况抛出通用错误
      throw new Error(`请求失败: ${response.status}`);
    }
  }

  const result: ApiResponse<T> = await response.json();

  if (!result.success) {
    throw new Error(result.error || 'Unknown error');
  }

  return result.data;
}

// ===================== API 函数 =====================

/**
 * 获取语料库列表
 */
export async function fetchCorpora(params?: {
  source_lang?: string;
  target_lang?: string;
  domain?: string;
  source_type?: string;
  is_public?: boolean;
  page?: number;
  limit?: number;
}): Promise<CorpusListResponse> {
  const searchParams = new URLSearchParams();

  if (params) {
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined && value !== null) {
        searchParams.append(key, String(value));
      }
    });
  }

  const query = searchParams.toString();
  const endpoint = `/corpus${query ? `?${query}` : ''}`;

  return apiRequest<CorpusListResponse>(endpoint);
}

/**
 * 获取单个语料库详情
 */
export async function fetchCorpusDetail(corpusId: number): Promise<CorpusItem> {
  const endpoint = `/corpus/${corpusId}`;
  return apiRequest<CorpusItem>(endpoint);
}

/**
 * 获取语料样本列表
 */
export async function fetchCorpusSamples(
  corpusId: number,
  params?: {
    page?: number;
    limit?: number;
  }
): Promise<SampleListResponse> {
  const searchParams = new URLSearchParams();

  if (params) {
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined && value !== null) {
        searchParams.append(key, String(value));
      }
    });
  }

  const query = searchParams.toString();
  const endpoint = `/corpus/${corpusId}/samples${query ? `?${query}` : ''}`;

  return apiRequest<SampleListResponse>(endpoint);
}

/**
 * 获取首页统计概览
 */
export async function fetchOverviewStats(): Promise<DashboardOverviewResponse> {
  return apiRequest<DashboardOverviewResponse>('/statistics/overview');
}

/**
 * 获取仪表盘详细统计
 */
export async function fetchDashboardStats(): Promise<DashboardStatsResponse> {
  return apiRequest<DashboardStatsResponse>('/dashboard/stats');
}

/**
 * KWIC 分析参数接口
 */
export interface KWICSearchParams {
  keyword: string;
  context_window?: 3 | 5 | 7;
  domain?: string;
  page?: number;
  limit?: number;
}

/**
 * KWIC 结果项接口
 */
export interface KWICResultItem {
  sentence_id: string;
  left_context: string;
  keyword: string;
  right_context: string;
  full_data?: {
    sentence_id: string;
    timestamp: string;
    platform: string;
    intent: string[];
    sentiment: string;
    business_scenario: string | null;
    source_text_zh: string;
    raw_text_ms: string;
    normalized_text_ms: string;
    english_loanwords: string[];
  };
}

/**
 * KWIC 响应接口
 */
export interface KWICResponse {
  items: KWICResultItem[];
  total: number;
  keyword: string;
  page: number;
  limit: number;
  domain?: string;
}

/**
 * 获取 KWIC 分析结果
 */
export async function fetchKWICAnalysis(
  corpusId: number,
  params: KWICSearchParams
): Promise<KWICResponse> {
  const searchParams = new URLSearchParams();

  if (params.keyword) searchParams.append('keyword', params.keyword);
  if (params.context_window) searchParams.append('context_window', String(params.context_window || 5));
  if (params.domain) searchParams.append('domain', params.domain);
  if (params.page) searchParams.append('page', String(params.page));
  if (params.limit) searchParams.append('limit', String(params.limit || 50));

  const query = searchParams.toString();
  const endpoint = `/corpus/${corpusId}/kwic/analysis?${query}`;

  return apiRequest<KWICResponse>(endpoint);
}

File: backend/api/schemas/corpus.py
"""
语料库相关的 Pydantic 模型
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Any
from datetime import datetime


# ===================== 嵌套结构 =====================

class BasicLayer(BaseModel):
    """基础层"""
    sentence_id: str
    timestamp: Optional[str] = None
    platform: Optional[str] = None


class LanguageLayer(BaseModel):
    """语言层"""
    source_text_zh: str = ""
    raw_text_ms: str = ""
    normalized_text_ms: str = ""
    english_loanwords: List[str] = []


class PragmaticLayer(BaseModel):
    """语用层"""
    intent: List[str] = []
    sentiment: str = "neutral"
    business_scenario: Optional[str] = None


class StyleLayer(BaseModel):
    """风格层"""
    style: Optional[str] = None
    contains_rojak: bool = False
    abbreviations_handled: dict = {}


class ScenarioTag(BaseModel):
    """场景标签"""
    label: str
    type: str


# ===================== 响应模型 =====================

class CorpusItem(BaseModel):
    """语料库列表项（兼容前端格式）"""
    id: int
    name: str
    sentences: str  # 格式化数字
    sTok: str       # 格式化 token 数
    tTok: str
    tags: List[ScenarioTag] = []
    domain: str = "general"
    source_type: str = "official"
    is_public: bool = True


class CorpusDetail(BaseModel):
    """语料库详情"""
    id: int
    name: str
    description: Optional[str] = None
    source_lang: str
    target_lang: str
    source_name: str
    target_name: str
    sentence_count: int
    source_token_count: int
    target_token_count: int
    domain: str
    source_type: str
    is_public: bool
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class CorpusSample(BaseModel):
    """语料样本（四层标注）"""
    basic_layer: BasicLayer
    language_layer: LanguageLayer
    pragmatic_layer: PragmaticLayer
    style_layer: StyleLayer


class SampleListResponse(BaseModel):
    """样本列表响应"""
    items: List[CorpusSample]
    total: int
    page: int = 1
    limit: int = 10


class CorpusListResponse(BaseModel):
    """语料库列表响应"""
    items: List[CorpusItem]
    total: int
    page: int = 1
    limit: int = 10


class CategoryStat(BaseModel):
    """分类统计"""
    name: str
    sentences: str
    percent: str


class DashboardOverviewResponse(BaseModel):
    """仪表盘概览响应"""
    corpus_count: str
    total_pairs: str
    language_count: str
    categories: List[CategoryStat]


class DashboardStatsResponse(BaseModel):
    """仪表盘详细统计响应"""
    # KPI 数据
    total_sentences: str
    sentences_growth: str
    avg_quality_score: str
    quality_target: str
    localization_value: str

    # 时间线数据
    timeline_data: List[dict]

    # 意图分布
    intents: List[dict]

    # 情感分布
    sentiment_distribution: dict

    # 渠道情感
    channel_sentiment: List[dict]

    # 样本数据
    samples: List[CorpusSample]

    # 借词统计
    loanwords: List[dict]

    # 标准化映射
    normalization_map: List[dict]

    # 风格分布
    style_distribution: List[dict]


# ===================== 请求模型 =====================

class CorpusSearchParams(BaseModel):
    """语料库搜索参数"""
    source_lang: Optional[str] = None
    target_lang: Optional[str] = None
    domain: Optional[str] = None
    source_type: Optional[str] = None
    is_public: Optional[bool] = None
    page: int = 1
    limit: int = 20

# ===================== KWIC 查询参数和响应 =====================
class KWICSearchParams(BaseModel):
    """KWIC 查询参数"""
    keyword: str = Field(..., description="目标词汇")
    context_window: int = Field(5, ge=1, le=10, description="前后词数")
    domain: Optional[str] = Field(None, description="业务域过滤")
    page: int = Field(1, ge=1, description="页码")
    limit: int = Field(50, ge=1, le=200, description="每页数量")

class KWICResultItem(BaseModel):
    """KWIC 结果项"""
    sentence_id: str = Field(..., description="句子ID")
    left_context: str = Field(..., description="左语境")
    keyword: str = Field(..., description="关键词")
    right_context: str = Field(..., description="右语境")
    full_data: Optional[dict] = Field(None, description="完整数据（可选）")

class KWICResponse(BaseModel):
    """KWIC 响应"""
    items: List[KWICResultItem] = Field(..., description="结果列表")
    total: int = Field(..., description="总结果数")
    keyword: str = Field(..., description="搜索关键词")
    page: int = Field(..., description="当前页码")
    limit: int = Field(..., description="每页数量")
    domain: Optional[str] = Field(None, description="过滤域")

class FrequencyWordItem(BaseModel):
    word: str = Field(..., description="词汇")
    count: int = Field(..., description="频次")
    percent: float = Field(..., description="占比")
    pos: str = Field(..., description="词性")

class CorpusFrequencyStatsResponse(BaseModel):
    total_words: int = Field(..., description="总词数")
    unique_words: int = Field(..., description="去重词数")
    pos_distribution: dict = Field(..., description="词性分布比例")
    frequency_data: List[FrequencyWordItem] = Field(..., description="高频词列表")
